server/core/src/physPath.cpp

```diff
@@ -1,47 +1,153 @@
+#include "rodsDef.h"
+#include "rodsDef.h"
+#include "rodsPath.h"
+#include "rsCollCreate.hpp"
 #include "rsDataObjClose.hpp"
-#include "genQuery.h"
-#include "phyBundleColl.h"
-#include "collCreate.h"
 #include "rsFileChksum.hpp"
-#include "rsModDataObjMeta.hpp"
+#include "rsFileCreate.hpp"
 #include "rsFileRename.hpp"
 #include "rsGenQuery.hpp"
-#include "rsCollCreate.hpp"
+#include "rsGlobalExtern.hpp"
+#include "rsModDataObjMeta.hpp"
 #include "rsObjStat.hpp"
-#include "irods_resource_backport.hpp"
+
+#include "irods_at_scope_exit.hpp"
+#include "irods_get_full_path_for_config_file.hpp"
 #include "irods_hierarchy_parser.hpp"
-#include "irods_stacktrace.hpp"
-#include "irods_server_properties.hpp"
 #include "irods_log.hpp"
-#include "irods_get_full_path_for_config_file.hpp"
 #include "irods_random.hpp"
+#include "irods_resource_backport.hpp"
+#include "irods_server_properties.hpp"
+#include "irods_stacktrace.hpp"
+#include "replica_proxy.hpp"
 
 #include <fmt/format.h>
 
 #include <unistd.h> // JMC - backport 4598
 #include <fcntl.h> // JMC - backport 4598
 
-int getLeafRescPathName( const std::string& _resc_hier, std::string& _ret_string );
+int getLeafRescPathName(const std::string& _resc_hier, std::string& _ret_string);
 
-int
-chkAndHandleOrphanFile( rsComm_t *rsComm, char* objPath, char* rescHIer, char *filePath,
-                        const char *_resc_name, int replStatus );
+int chkAndHandleOrphanFile(
+    rsComm_t *rsComm,
+    char* objPath,
+    char* rescHier,
+    char *filePath,
+    const char *_resc_name,
+    int replStatus);
+
+namespace
+{
+    namespace ir = irods::experimental::replica;
+} // anonymous namespace
+
+namespace irods
+{
+    auto create_physical_file_for_replica(
+        RsComm&      _comm,
+        DataObjInp&  _inp,
+        DataObjInfo& _replica_info) -> int
+    {
+        const int chkType = getchkPathPerm(&_comm, &_inp, &_replica_info);
+        if (DISALLOW_PATH_REG == chkType) {
+            return PATH_REG_NOT_ALLOWED;
+        }
+
+        auto r = ir::make_replica_proxy(_replica_info);
+
+        fileCreateInp_t fileCreateInp{};
+        rstrcpy(fileCreateInp.resc_name_,    r.resource().data(),       MAX_NAME_LEN);
+        rstrcpy(fileCreateInp.resc_hier_,    r.hierarchy().data(),      MAX_NAME_LEN);
+        rstrcpy(fileCreateInp.objPath,       r.logical_path().data(),   MAX_NAME_LEN);
+        rstrcpy(fileCreateInp.fileName,      r.physical_path().data(),  MAX_NAME_LEN);
+        rstrcpy(fileCreateInp.in_pdmo,       r.in_pdmo().data(),        MAX_NAME_LEN );
+        fileCreateInp.mode = getFileMode(&_inp);
+
+        const auto clear_cond_input = irods::at_scope_exit{[&fileCreateInp] { clearKeyVal(&fileCreateInp.condInput); }};
+        copyKeyVal(&_replica_info.condInput, &fileCreateInp.condInput);
+
+        if (NO_CHK_PATH_PERM == chkType) {
+            fileCreateInp.otherFlags |= NO_CHK_PERM_FLAG;
+        }
+
+        int fd = 0;
+        constexpr auto maximum_retries = 100;
```

Comment:
is this a defensible number?   from other code in the world?  best practice?    why 100?

Reply:
This is from the beforetimes. Just captured it in a variable. It seems a little high to me (usually just needs 1 retry because there's a leftover directory or something there) but I'm willing to change it to whatever we decide.

Reply:
right, 5 seems way plenty.   but also, i can't really think of a downside of it being larger.  just uses a tad more time before a failure message occurs.

Reply:
Okay. If it's all the same to us, then, I think I'd like to leave it as is.

Reply:
yeps.